### PR TITLE
implement iterator pattern

### PR DESCRIPTION
**Pull request by araddon**
Implement the Iterator pattern to make it easier to deal with large file lists

### CODE REVIEW

File: store.go [hunk 1]
```diff
@@ -1,3 +1,6 @@
+// Cloud Storage interface to make Local, Google, s3 file storage
```
Comment (timkaye11) [inline store.go]: if you put this in a `doc.go` it'll look nicer in godoc

Comment (araddon) [inline store.go]: It does pull out from any file:

https://github.com/araddon/qlbridge/blob/master/datasource/memdb/db.go

https://godoc.org/github.com/araddon/qlbridge/datasource/memdb

Or maybe i don't understand, is there a good example?

File: gcsstore.go [hunk 2]
```diff
@@ -207,6 +193,34 @@ func (g *GcsFS) Delete(obj string) error {
...
+
+	for {
+		o, err := it.iter.Next()
+		if err == nil {
+			return newObjectFromGcs(it.g, o), nil
+		} else if err == iterator.Done {
+			return nil, err
+		}
+		lasterr = err
+		if retryCt < 5 {
```
Comment (epsniff) [inline gcsstore.go]: 1. at LN 133 we should memorize the ctx passed in.
2. here we should check if the it.ctx.Done() has been closed.

File: gcsstore.go [hunk 3]
```diff
@@ -207,6 +193,34 @@ func (g *GcsFS) Delete(obj string) error {
...
+
+func (it *GcsObjectIterator) Next() (Object, error) {
+	var lasterr error = nil
+	retryCt := 0
+
+	for {
+		o, err := it.iter.Next()
+		if err == nil {
+			return newObjectFromGcs(it.g, o), nil
+		} else if err == iterator.Done {
```
Comment (epsniff) [inline gcsstore.go]: I would change this to be:
```
} else if err == iterator.Done || err == context.Canceled || context.DeadlineExceeded {
```

File: store.go (hunk 1 above)
Comment (timkaye11) [inline store.go]: oh whoops you're right, it does pull it out from any file

**Pull request merged**